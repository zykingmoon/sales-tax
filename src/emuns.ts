export enum ProductCategories {
    BOOK,
    FOOD,
    MEDICAL,
    MUSIC,
    PERFUME
  }
export interface ProductProps {
    id: number;
    isImported: boolean;
    name: string;
    price: number;
    category: number;
    amount: number;
}

export interface PriceProps {
    id: number;
    name: string;
    amount: number;
    total: string;
}


export const salesTaxExceptionCategoris: number[] = [ProductCategories['BOOK'], ProductCategories['FOOD'], ProductCategories['MEDICAL']]
export const salesTaxRate: number = 0.1;
export const importDutyRate: number = 0.05;