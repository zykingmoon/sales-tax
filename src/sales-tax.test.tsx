/* eslint-disable testing-library/await-async-query */

import {create, act} from 'react-test-renderer';
import PriceDetails from "./sales-tax"
import { ProductCategories, ProductProps } from './emuns';


const input1: ProductProps[] = [{
  id: 111,
  isImported: false,
  amount: 1,
  name: 'book',
  price: 12.49,
  category: ProductCategories['BOOK']
}, {
  id: 112,
  isImported: false,
  amount: 1,
  name: 'music CD',
  price: 14.99,
  category: ProductCategories['MUSIC']
}, {
  id: 113,
  isImported: false,
  amount: 1,
  name: 'chocolate bar',
  price: 0.85,
  category: ProductCategories['FOOD']
}];

const input2: ProductProps[] = [{
  id: 221,
  isImported: true,
  amount: 1,
  name: 'imported box of chocolates',
  price: 10.00,
  category: ProductCategories['FOOD']
}, {
  id: 222,
  isImported: true,
  amount: 1,
  name: 'imported bottle of perfume',
  price: 47.50,
  category: ProductCategories['PERFUME']
}];

const input3: ProductProps[] = [{
  id: 331,
  isImported: true,
  amount: 1,
  name: 'imported bottle of perfume',
  price: 27.99,
  category: ProductCategories['PERFUME']
}, {
  id: 332,
  isImported: false,
  amount: 1,
  name: 'bottle of perfume',
  price: 18.99,
  category: ProductCategories['PERFUME']
}, {
  id: 333,
  isImported: false,
  amount: 1,
  name: 'packet of headache pills',
  price: 9.75,
  category: ProductCategories['MEDICAL']
}, {
  id: 334,
  isImported: true,
  amount: 1,
  name: 'imported box of chocolates',
  price: 11.25,
  category: ProductCategories['FOOD']
}];

it('changes the class when hovered', () => {

  let root; 
  act(() => {
    root = create(<PriceDetails goods={input1} />)
  });

  // 对根元素进行断言
  expect(root.toJSON()).toMatchSnapshot();

  const testInstance = root.root;
  
  // 
  let priceItems = testInstance.findAllByProps({className: "priceItem"});
  expect(priceItems[0].findByProps({className: "itemTitle"}).children).toEqual(['1', ' ', 'book', ': ']);
  expect(priceItems[0].findByProps({className: "itemPrice"}).children).toEqual(['12.49']);
  expect(priceItems[1].findByProps({className: "itemTitle"}).children).toEqual(['1', ' ', 'music CD', ': ']);
  expect(priceItems[1].findByProps({className: "itemPrice"}).children).toEqual(['16.49']);
  expect(priceItems[2].findByProps({className: "itemTitle"}).children).toEqual(['1', ' ', 'chocolate bar', ': ']);
  expect(priceItems[2].findByProps({className: "itemPrice"}).children).toEqual(['0.85']);

  expect(testInstance.findByProps({className: "priceTax"}).findByProps({className: "itemPrice"}).children).toEqual(['1.50']);
  expect(testInstance.findByProps({className: "priceTotal"}).findByProps({className: "itemPrice"}).children).toEqual(['29.83']);

  // 更新 props
  act(() => {
    root.update(<PriceDetails goods={input2} />);
  })
  priceItems = testInstance.findAllByProps({className: "priceItem"});
  expect(priceItems[0].findByProps({className: "itemTitle"}).children).toEqual(['1', ' ', 'imported box of chocolates', ': ']);
  expect(priceItems[0].findByProps({className: "itemPrice"}).children).toEqual(['10.50']);
  expect(priceItems[1].findByProps({className: "itemTitle"}).children).toEqual(['1', ' ', 'imported bottle of perfume', ': ']);
  expect(priceItems[1].findByProps({className: "itemPrice"}).children).toEqual(['54.65']);

  expect(testInstance.findByProps({className: "priceTax"}).findByProps({className: "itemPrice"}).children).toEqual(['7.65']);
  expect(testInstance.findByProps({className: "priceTotal"}).findByProps({className: "itemPrice"}).children).toEqual(['65.15']);


  // 更新 props
  act(() => {
    root.update(<PriceDetails goods={input3} />);
  })
  priceItems = testInstance.findAllByProps({className: "priceItem"});
  expect(priceItems[0].findByProps({className: "itemTitle"}).children).toEqual(['1', ' ', 'imported bottle of perfume', ': ']);
  expect(priceItems[0].findByProps({className: "itemPrice"}).children).toEqual(['32.19']);
  expect(priceItems[1].findByProps({className: "itemTitle"}).children).toEqual(['1', ' ', 'bottle of perfume', ': ']);
  expect(priceItems[1].findByProps({className: "itemPrice"}).children).toEqual(['20.89']);
  expect(priceItems[2].findByProps({className: "itemTitle"}).children).toEqual(['1', ' ', 'packet of headache pills', ': ']);
  expect(priceItems[2].findByProps({className: "itemPrice"}).children).toEqual(['9.75']);
  expect(priceItems[3].findByProps({className: "itemTitle"}).children).toEqual(['1', ' ', 'imported box of chocolates', ': ']);
  expect(priceItems[3].findByProps({className: "itemPrice"}).children).toEqual(['11.85']);

  expect(testInstance.findByProps({className: "priceTax"}).findByProps({className: "itemPrice"}).children).toEqual(['6.70']);
  expect(testInstance.findByProps({className: "priceTotal"}).findByProps({className: "itemPrice"}).children).toEqual(['74.68']);

});

