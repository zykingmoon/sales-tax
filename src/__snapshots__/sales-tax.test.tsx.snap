// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes the class when hovered 1`] = `
<div
  className="priceDetails"
>
  <div
    className="priceItem"
  >
    <div
      className="itemTitle"
    >
      1
       
      book
      : 
    </div>
    <div
      className="itemPrice"
    >
      12.49
    </div>
  </div>
  <div
    className="priceItem"
  >
    <div
      className="itemTitle"
    >
      1
       
      music CD
      : 
    </div>
    <div
      className="itemPrice"
    >
      16.49
    </div>
  </div>
  <div
    className="priceItem"
  >
    <div
      className="itemTitle"
    >
      1
       
      chocolate bar
      : 
    </div>
    <div
      className="itemPrice"
    >
      0.85
    </div>
  </div>
  <div
    className="priceTax"
  >
    <div
      className="itemTitle"
    >
      Sales Taxes: 
    </div>
    <div
      className="itemPrice"
    >
      1.50
    </div>
  </div>
  <div
    className="priceTotal"
  >
    <div
      className="itemTitle"
    >
      Total: 
    </div>
    <div
      className="itemPrice"
    >
      29.83
    </div>
  </div>
</div>
`;
